src/geoip_cache.rs

```diff
@@ -0,0 +1,105 @@
+use serde::{Deserialize, Serialize};
+use std::net::IpAddr;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::time::SystemTime;
+use std::{collections::HashMap, fs, io, time::Duration};
+use tokio::sync::RwLock;
+
+use ziggurat_core_geoip::geoip::{GeoIPInfo, GeoIPService};
+
+#[derive(Clone, Serialize, Deserialize)]
+struct CacheEntry {
+    pub last_updated: SystemTime,
+    pub entry: GeoIPInfo,
+}
+
+#[derive(Default, Clone, Serialize, Deserialize)]
+struct GeoCache {
+    pub entries: HashMap<IpAddr, CacheEntry>,
+}
+
+/// GeoIP cache responsible for getting and caching results.
+pub struct GeoIPCache {
+    /// Available providers and their configuration.
+    providers: Vec<Box<dyn GeoIPService>>,
+    /// Path to the cache file.
+    cache_file: PathBuf,
+    /// Cache entries.
+    cache: Arc<RwLock<GeoCache>>,
+}
+
+impl GeoIPCache {
+    /// Create a new GeoIP cache.
+    pub fn new(cache_file: &Path) -> Self {
+        Self {
+            providers: Vec::new(),
+            cache_file: cache_file.to_owned(),
+            cache: Arc::new(RwLock::new(GeoCache::default())),
+        }
+    }
+
+    /// Add a new provider to the list of providers. The providers will be called in the order they
+    /// are added.
+    pub fn add_provider(&mut self, provider: Box<dyn GeoIPService>) {
+        self.providers.push(provider);
+    }
+
+    /// Load the cache from the file.
+    pub async fn load(&self) -> Result<(), io::Error> {
+        let cache_string = fs::read_to_string(&self.cache_file)?;
+
+        let mut cache = self.cache.write().await;
+        cache.entries = serde_json::from_str(&cache_string).unwrap();
+        Ok(())
+    }
+
+    /// Save the cache to the file.
+    pub async fn save(&self) -> Result<(), io::Error> {
+        let cache = self.cache.read().await;
+        let cache_string = serde_json::to_string(&cache.entries).unwrap();
+        fs::write(&self.cache_file, cache_string)
+    }
+
+    /// Function look in cache and if not found, it will call the providers to fetch new data and
+    /// store it into cache.
+    pub async fn lookup(&self, ip: IpAddr) -> Option<GeoIPInfo> {
+        let mut remove_entry = false;
+
+        {
+            let cache = self.cache.read().await;
```

Comment:
This whole block (actually from lines 69-85, it's all cache related) should be a function - something like `self.check_cache(ip).await`

such function should also take care of clearing old entries

Reply:
Done.